File: firebase/functions/lib/events/event_ended.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart'
    as admin_interop;
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/email_templates.dart';
import '../utils/infra/firestore_utils.dart';
import '../utils/notifications_utils.dart';
import '../utils/subscription_plan_util.dart';
import 'live_meetings/agora_api.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/events/live_meetings/live_meeting.dart';
import 'package:data_models/community/community.dart';

/// This function handles events after event ends
class EventEnded extends OnCallMethod<EventEndedRequest> {
  static const String kEventEndedApi = 'eventEnded';
  NotificationsUtils notificationsUtils;
  AgoraUtils agoraUtils;

  EventEnded({NotificationsUtils? notificationsUtils, AgoraUtils? agoraUtils})
      : notificationsUtils = notificationsUtils ?? NotificationsUtils(),
        agoraUtils = agoraUtils ?? AgoraUtils(),
        super(
          kEventEndedApi,
          (jsonMap) => EventEndedRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    EventEndedRequest request,
    CallableContext context,
  ) async {
    final participant = await firestoreUtils.getFirestoreObject(
      path: '${request.eventPath}/event-participants/${context.authUid}',
      constructor: (map) => Participant.fromJson(map),
    );

    // Check that user is an active participant
    if (participant.status != ParticipantStatus.active) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final event = await firestoreUtils.getFirestoreObject(
      path: request.eventPath,
      constructor: (map) => Event.fromJson(map),
    );

    // Stop main room recording if one is active.
    final liveMeetingPath = '${request.eventPath}/live-meetings/${event.id}';
    try {
      final liveMeeting = await firestoreUtils.getFirestoreObject(
        path: liveMeetingPath,
        constructor: (map) => LiveMeeting.fromJson(map),
      );
      if (liveMeeting.recordingSessionId != null) {
        await agoraUtils.stopRoom(sessionId: liveMeeting.recordingSessionId!);
      }
    } catch (e) {
      // Do not block event-ended flow on recording stop failure.
      print('Error stopping main room recording on event end: $e');
    }

    // Stop all breakout room recordings.
    // Structure: {liveMeetingPath}/breakout-room-sessions/{sessionId}/breakout-rooms/{roomId}
    try {
      final breakoutSessionDocs = await firestore
          .collection('$liveMeetingPath/breakout-room-sessions')
          .get();
      for (final sessionDoc in breakoutSessionDocs.documents) {
        final breakoutRoomDocs = await firestore
            .collection('${sessionDoc.reference.path}/breakout-rooms')
            .get();
        for (final roomDoc in breakoutRoomDocs.documents) {
          final breakoutRoom = BreakoutRoom.fromJson(
            firestoreUtils.fromFirestoreJson(roomDoc.data.toMap()),
          );
          if (breakoutRoom.recordingSessionId != null) {
            try {
              await agoraUtils.stopRoom(
                  sessionId: breakoutRoom.recordingSessionId!);
            } catch (e) {
              print(
                  'Error stopping breakout recording ${breakoutRoom.recordingSessionId}: $e');
            }
          }
        }
      }
    } catch (e) {
      print('Error stopping breakout room recordings on event end: $e');
    }

    final capabilities =
        await subscriptionPlanUtil.calculateCapabilities(event.communityId);
    final hasPrePost = capabilities.hasPrePost ?? false;

    await notificationsUtils.sendEventEndedEmail(
      event: event,
      communityId: event.communityId,
      userIds: [context.authUid!],
      emailType: EventEmailType.ended,
      generateMessage: (Community community, admin_interop.UserRecord user) =>
          SendGridEmailMessage(
        subject: 'Thanks for joining',
        html: generateEventEndedContent(
          header: 'Thanks for joining ${event.title}!',
          community: community,
          userRecord: user,
          event: event,
          allowPrePost: hasPrePost,
        ),
      ),
    );
  }
}

File: firebase/functions/src/events/event_ended.ts
import * as functions from 'firebase-functions';
import { OnCallMethod } from '../on_call_function';
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils';
import { notificationsUtils } from '../utils/notifications_utils';
import { subscriptionPlanUtil } from '../utils/subscription_plan_util';
import { generateEventEndedContent } from '../utils/email_templates';
import { AgoraUtils } from './live_meetings/agora_api';
import { Event, Participant, ParticipantStatus, BreakoutRoom, EventEmailType } from '../types';

interface EventEndedRequest {
  eventPath: string;
}

export class EventEnded extends OnCallMethod<EventEndedRequest> {
  private agoraUtils: AgoraUtils;

  constructor() {
    super('eventEnded', (jsonMap) => jsonMap as EventEndedRequest);
    this.agoraUtils = new AgoraUtils();
  }

  async action(request: EventEndedRequest, context: functions.https.CallableContext): Promise<void> {
    const participant: Participant = await firestoreUtils.getFirestoreObject({
      path: `${request.eventPath}/event-participants/${context.auth?.uid}`,
      constructor: (map) => map as unknown as Participant,
    });

    if (participant.status !== ParticipantStatus.active) {
      throw new functions.https.HttpsError('failed-precondition', 'unauthorized');
    }

    const event: Event = await firestoreUtils.getFirestoreObject({
      path: request.eventPath,
      constructor: (map) => map as unknown as Event,
    });

    // Stop main room recording
    const liveMeetingPath = `${request.eventPath}/live-meetings/${event.id}`;
    try {
      const liveMeeting = await firestoreUtils.getFirestoreObject({
        path: liveMeetingPath,
        constructor: (map) => map as unknown as any,
      });
      if (liveMeeting.recordingSessionId) {
        await this.agoraUtils.stopRoom({ sessionId: liveMeeting.recordingSessionId });
      }
    } catch (e) {
      console.error('Error stopping main room recording:', e);
    }

    // Stop breakout recordings
    try {
      const breakoutSessionDocs = await firestore.collection(`${liveMeetingPath}/breakout-room-sessions`).get();
      for (const sessionDoc of breakoutSessionDocs.docs) {
        const breakoutRoomDocs = await firestore.collection(`${sessionDoc.ref.path}/breakout-rooms`).get();
        for (const roomDoc of breakoutRoomDocs.docs) {
          const breakoutRoom: BreakoutRoom = firestoreUtils.fromFirestoreJson(roomDoc.data()) as unknown as BreakoutRoom;
          if (breakoutRoom.recordingSessionId) {
            try {
              await this.agoraUtils.stopRoom({ sessionId: breakoutRoom.recordingSessionId });
            } catch (e) {
              console.error('Error stopping breakout recording:', e);
            }
          }
        }
      }
    } catch (e) {
      console.error('Error stopping breakout recordings:', e);
    }

    const capabilities = await subscriptionPlanUtil.calculateCapabilities(event.communityId!);
    const hasPrePost = capabilities.hasPrePost ?? false;

    await notificationsUtils.sendEventEndedEmail({
      event,
      communityId: event.communityId!,
      userIds: [context.auth!.uid],
      emailType: EventEmailType.ended,
      generateMessage: (community, user) => ({
        subject: 'Thanks for joining',
        html: generateEventEndedContent({
          header: `Thanks for joining ${event.title}!`,
          community,
          userRecord: user,
          event,
          allowPrePost: hasPrePost,
        }),
      }),
    });
  }
}
